Proof Scripts in JML

mattulbrich opened the pull request:
## Related Issue

For quite some time, we have a prototype for JML proof scripts which we finally want to bring to the mater branch.

## Intended Change

Proof scripts can be written into JML comments and can be obeyed during automatic verification.
The major benefit is that the localisation of the proof node to which a proof script should be applied is natural.
The other benefit is that one can use JML expressions and thus does not have to leave the realm of JML to do script-guided proofs.

This relies on #3587 and needs #3654 merged.

## Plan

* [x] Refine the syntax
* [x] Adapt lexer and parser
* [x] Reimport the implementation that was available on an older branch
* [x] Make it more convenient
* [x] Add examples
* [x] Code cleanup
* [x] Document the changes

## Type of pull request

- New feature (non-breaking change which adds functionality)
- There are changes to the (Java) code
- There are changes to the taclet rule base (possibly small adaptations)

## Ensuring quality

WIP:
- I made sure that introduced/changed code is well documented (javadoc and inline comments). :heavy_check_mark: 
- I made sure that new/changed end-user features are well documented (https://github.com/KeYProject/key-docs). :heavy_check_mark: 
- I added new test case(s) for new functionality. :heavy_check_mark: 
- I have tested the feature as follows: ... examples and test cases
- I have checked that runtime performance has not deteriorated. not for non-script proofs.

## Additional information and contact(s)

Planned to be done Mid 09/25

The contributions within this pull request are licensed under GPLv2 (only) for inclusion in KeY.

mattulbrich committed:
added missing files for script aware macros

(cherry-picked from earlier attempt to implement JML proof scripts)

mattulbrich committed:
advancing the immutable list interface

mattulbrich committed:
advancing for scripts from JML

mattulbrich committed:
slight adaptation of the macro used for scripting

mattulbrich committed:
more script commands

mattulbrich committed:
improving/correcting script commands

mattulbrich committed:
settings for the auto command

mattulbrich committed:
propagating strategy settings to the strategy to be used

(refers to code available in SetCommand)

mattulbrich committed:
proof scripts: adding a CHEAT command

mattulbrich committed:
updating a few of the script commands

mattulbrich committed:
introducing the script-aware prep macro

mattulbrich committed:
issue dialog: skip spaces for squiggly lines

mattulbrich committed:
value injector: towards reporting unknown arguments

mattulbrich committed:
improved script error reporting

mattulbrich committed:
documentation for script commands

mattulbrich committed:
better error messaging in ScriptLineParser

mattulbrich committed:
implement a recall mechanism

mattulbrich committed:
allow let commands without "@"

mattulbrich committed:
a formula parameter for the expand command

mattulbrich committed:
introducing hole placeholders for terms and for formulas

# Conflicts:
#	key.core/src/main/resources/de/uka/ilkd/key/proof/rules/javaHeader.key

mattulbrich committed:
introducing sort::FOCUS for usage in scripts

mattulbrich committed:
introducing sort::ELLIP for ellipsis search patterns

mattulbrich committed:
add "assumes" parameter to rule command

not functional yet after cherry-picking

mattulbrich committed:
introducing "auto add_*" to scripts.

mattulbrich committed:
introducing witness command

WolframPfeifer committed:
allow terms with holes in expand command

adapted to new branch by MU

WolframPfeifer committed:
added convenience macro to close all NPE and Out-of-bounds branches

mattulbrich committed:
first steps to allow scripts in jml

(cherry-picked from earlier attempt to implement JML proof scripts)

# Conflicts:
#	key.core/src/main/java/de/uka/ilkd/key/java/Recoder2KeYConverter.java
#	key.core/src/main/java/de/uka/ilkd/key/java/recoderext/JMLTransformer.java
#	key.core/src/main/java/de/uka/ilkd/key/java/recoderext/JmlAssert.java
#	key.core/src/main/java/de/uka/ilkd/key/java/statement/JmlAssert.java
#	key.core/src/main/java/de/uka/ilkd/key/speclang/jml/pretranslation/TextualJMLAssertStatement.java
#	key.core/src/main/java/de/uka/ilkd/key/speclang/njml/TextualTranslator.java

# Conflicts:
#	key.core/src/main/antlr4/JmlParser.g4
#	key.core/src/main/java/de/uka/ilkd/key/java/Recoder2KeYConverter.java
#	key.core/src/main/java/de/uka/ilkd/key/java/ast/statement/JmlAssert.java
#	key.core/src/main/java/de/uka/ilkd/key/java/recoderext/JMLTransformer.java
#	key.core/src/main/java/de/uka/ilkd/key/java/recoderext/JmlAssert.java
#	key.core/src/main/java/de/uka/ilkd/key/speclang/njml/TextualTranslator.java

mattulbrich committed:
first working example of a proof script in Java code (albeit with JavaDL arguments)

(cherry-picked from earlier attempt to implement JML proof scripts)

mattulbrich committed:
more infrastructure for proof scripts

# Conflicts:
#	key.core/src/main/java/de/uka/ilkd/key/logic/Semisequent.java
#	key.core/src/main/java/de/uka/ilkd/key/scripts/RuleCommand.java
#	key.core/src/main/java/de/uka/ilkd/key/scripts/SetCommand.java
#	key.core/src/main/java/de/uka/ilkd/key/strategy/StrategyProperties.java

more infrastructure for proof scripts

(cherry-picked from earlier attempt to implement JML proof scripts)

more infrastructure for proof scripts

(cherry-picked from earlier attempt to implement JML proof scripts)

# Conflicts:
#	key.core/src/main/java/de/uka/ilkd/key/macros/ApplyScriptsMacro.java

mattulbrich committed:
adapting to new Script AST nodes

mattulbrich committed:
renaming AssertCommand to FailUnlessCommand

mattulbrich committed:
correcting the grammar for nested \by clauses

wadoon commented:
I have written an integer split command. 

An induction command is on the way with support for int, user-data types, bsum(?). 

Commit: 4644331608e051e73884ef2c11b66490d189fca5

mattulbrich committed:
improving the value injector

mattulbrich committed:
renaming a deprecated command "assert" --> "assertOpenGoals"

mattulbrich committed:
spotless

mattulbrich committed:
working version of scripts

mattulbrich committed:
Position info for scripts with url=null, run scripted goals before other goals, print position

manual cherry-pick

From 547ce291d230a59b3abf4596ccede2e6a64aed75 Mon Sep 17 00:00:00 2001
From: Julian Wiesler <wiesleju@gmail.com>
Date: Wed, 22 Feb 2023 13:57:21 +0100
Subject: [PATCH] Position info for scripts with url=null, run scripted goals
 before other goals, print position information

mattulbrich committed:
Use AutoPilotPrepareProofMacro instead of FinishSymbolicExecutionMacro since it splits a lot better

# Conflicts:
#	key.core/src/main/java/de/uka/ilkd/key/macros/ApplyScriptsMacro.java

mattulbrich committed:
working on improved script commands

manual cherry-picking

commit 886588af1c72cdbf86b7a624a4045c6b7fb6a4b2
Author: Mattias Ulbrich <ulbrich@kit.edu>
Date:   Sun Feb 5 13:14:29 2023 +0100

mattulbrich committed:
reorganisation of proof script commands

manually cherry-picked from old branch

mattulbrich commented:
The history of this PR was totally changed while I was working on it.
I will have to force-push my changes of the last few days.
In case it is still needed: e841235708a1099a7e0b74c9af68c3c7bf8d5369

mattulbrich committed:
spotlessing

after a lengthy cherry-pick session from an old branch

mattulbrich committed:
towards 'obtain' in JML scripts

mattulbrich committed:
update ProofScriptEngine's workflow

after discussion with Alexander

mattulbrich committed:
consolidating, introducing test cases for jml scripts

mattulbrich committed:
first working obtain scripts

mattulbrich committed:
pimping the document generation for proof script commands

mattulbrich committed:
lots of script documentation

mattulbrich committed:
fixing a bug regarding proof script application

mattulbrich committed:
repairing some weird self variable treatment in SpecStatement

mattulbrich committed:
The Boyer-Moore example now runs on scripts

mattulbrich committed:
smaller fixes in proof script treatment

mattulbrich committed:
improved document generation

mattulbrich committed:
more proof script power, update inlining.

mattulbrich committed:
making sure that cuts work with boolean terms instead of formulas.

mattulbrich committed:
better symbex-only machine

mattulbrich committed:
Allowing "auto only: ..."

mattulbrich committed:
quicksort example with JML scripts (partially)

mattulbrich committed:
enabling the quicksort example with JML proof scripts!

mattulbrich committed:
limiting the symbex only macro a bit

added test cases

mattulbrich committed:
adapting the cherry-picked commits

mattulbrich committed:
allowing switching off rules in auto

from 53286f761f547ae662e6275fb0693b92f2c502ba

mattulbrich committed:
a more robust term-with-holes implementation

mattulbrich committed:
improved treatment of terms with holes

mattulbrich committed:
introduce match identifiers with leading '?'

mattulbrich committed:
improving sophisticated term matching

there are failing cases ...

mattulbrich committed:
accomodating "focus inside find"

which is algorithmically not trivial!

mattulbrich committed:
bugfixing matching with program variables

mattulbrich committed:
updating the focus rule

mattulbrich committed:
bug fixes in script engine

mattulbrich committed:
repairing JML script tests

mattulbrich committed:
improving symbex only macro

mattulbrich committed:
adding an infinity catch on "throw null" in symb ex.

mattulbrich committed:
making the assume command use a local taclet.

mattulbrich committed:
correcting conversion in JML scripts

mattulbrich committed:
adapting taclets.old.txt to changed throws treatment

mattulbrich committed:
Repaired a merge test case

mattulbrich committed:
missing sameUpdateLevel for tryCatchThrow

wadoon committed:
fixing after rebase

mattulbrich committed:
repairing the nested array jml script test case

mattulbrich committed:
small issues in the script engine to make BM proofs possible

mattulbrich commented:
I believe I fixed the functional issues that were left. It remains:
* Documentation
* Rebasing or merging with main resolving the above conflicts [which are not 100% clear]

There is a chance that this can be done by the end of the month.

wadoon committed:
Merge branch 'main' into fork/mattulbrich/jmlScripts

# By Richard Bubel (29) and others
# Via GitHub (25) and others
* main: (76 commits)
  Show a proof macro's aggregate statistics in the status line
  Don't crash reloading a recent file with no stored profile
  Robustly read numeric settings stored as Integer or Long
  remove proof pruning changes (were not working), update task bar button
  Fix 'choose contract' in parser
  Remove missing semi tests
  Make semicolon after rule optional
  Offer the classic taclet-match dialog as a migration fallback
  Improve error message when typing in assumes sequent
  spotless
  Bump the gradle-deps group across 1 directory with 8 updates
  Rework the taclet-match dialog and retire the old one
  Add GUI-independent text and assumes-input logic with unit tests
  Add a side-effect-free result-sequent preview to the taclet executor
  Index applied rule apps by focus-term fingerprint for O(1) duplicate detection
  Remove larger problems from tests due to memory limitations on CI
  Add update-simplification performance regression tests to runAllProofs
  Test targeted \dropEffectlessElementaries
  Speed up update simplification: targeted \dropEffectlessElementaries
  Add a done() early-exit short-circuit to JavaASTWalker
  ...

# Conflicts:
#	key.core/src/main/antlr4/KeYLexer.g4
#	key.core/src/main/antlr4/KeYParser.g4
#	key.core/src/main/java/de/uka/ilkd/key/nparser/ParsingFacade.java
#	key.core/src/main/java/de/uka/ilkd/key/nparser/builder/DefaultBuilder.java
#	key.core/src/main/java/de/uka/ilkd/key/scripts/LetCommand.java
#	key.core/src/main/java/de/uka/ilkd/key/scripts/ProofScriptEngine.java
#	key.core/src/main/java/de/uka/ilkd/key/scripts/ScriptCommandAst.java
#	key.ui/src/main/java/de/uka/ilkd/key/gui/IssueDialog.java
#	key.util/src/main/java/org/key_project/util/collection/ImmutableSet.java

wadoon committed:
Merge branch 'main' into jmlScripts

wadoon committed:
fix LetCommand after merge

wadoon committed:
remove semantics guard allowMatchId

wadoon committed:
repair antlr import warnings

mattulbrich committed:
Adding JavaDoc to some relevant proof script related classes

mattulbrich committed:
Javadoc and Documentation annotations for commands

mattulbrich committed:
Merge remote-tracking branch 'origin/main' into jmlScripts

# Conflicts:
#	key.core/build.gradle

mattulbrich committed:
spotless, small adaptations

mattulbrich commented:
It seems there is one open test case that has little to do with scripts but with error reporting:

The test case is `(true & true` (missing closing paren), where the error message was:

    ')' (closing parenthesis) expected, but found '}'

but now reads:

    one of '\if', '\ifEx', 'true', 'false', UTF_EMPTY, '(' (opening parenthesis), ... expected, but found '}'

which is obviously wrong. It turns out the parser bails out at the `(` and reports what it expects there and says that it found '}' instead. So these two locations seem to refer to different places in the parsed text.

The last commit that reworked this error messaging is rather recent:  a3480186e9e9e217e820a15636169bdae551eb12

@unp1, perhaps you can look into it?

mattulbrich committed:
deactivating a parser rule which seems unnecessary

(if necessary, then we should refactor the grammar here ...)

mattulbrich committed:
Merge branch 'main' into jmlScripts

mattulbrich commented:
> @unp1, perhaps you can look into it?

While this case shows that something is still amiss with the error reporting, this particular instance could be resolved by commenting out one line ... but considering a parenthesised term as a literal was not ideal anyway.

So: no immediate action needed here. Sorry if I caused inconvenience.

unp1 commented:
> > @unp1, perhaps you can look into it?
> 
> While this case shows that something is still amiss with the error reporting, this particular instance could be resolved by commenting out one line ... but considering a parenthesised term as a literal was not ideal anyway.
> 
> So: no immediate action needed here. Sorry if I caused inconvenience.

no problem. Sorry for not answwering earlier, was preparing for a journal deadline today and did not see it until now.

mattulbrich commented:
<img width="614" height="42" alt="grafik" src="https://github.com/user-attachments/assets/2aa679d7-a5d1-4c17-9962-4d84872c32c8" />
<img width="728" height="31" alt="grafik" src="https://github.com/user-attachments/assets/462f287c-eb73-4a37-b442-880ecd8ad2ca" />

I have no idea why the Windows test cases fail. I even used my Windows PC at home to run the tests: And they succeed. ...

unp1 commented:
> I have no idea why the Windows test cases fail. I even used my Windows PC at home to run the tests: And they succeed. ...

Apparently (at least C's analysis) line endings on Windows. It suggests two fixes:

- ScriptLineParser so that users do not run into the problem:
   replace "case '\r' -> { }"  with "case '\r' -> { continue; }" to ignore '\r' as otherwise it ends up in 
   if (state != State.IN_COMMENT) {
    cmdBuilder.append((char) c);   // <-- runs for '\r' too
}

and which I am not such a fan of:

- in .gitattributes (to make tests more robust)
+# The embedded JML proof scripts in these inputs must parse identically on all
+# platforms. Force LF on checkout so Windows does not introduce CR characters
+# that the proof-script tokenizer (ScriptLineParser) is sensitive to.
+*.java text eol=lf

mattulbrich committed:
TEMPORARY COMMIT TO INVESTIGATE TEST CASE FAILURE

mattulbrich committed:
TEMPORARY COMMIT TO INVESTIGATE TEST CASE FAILURE

mattulbrich committed:
TEMPORARY COMMIT TO INVESTIGATE TEST CASE FAILURE

mattulbrich committed:
Merge commit '7828992f4c97e7d893bb0bdbdd5f7981e6501799' into jmlScripts

mattulbrich commented:
> > I have no idea why the Windows test cases fail. I even used my Windows PC at home to run the tests: And they succeed. ...

Thanks for investigating, Richard @unp1 

> Apparently (at least C's analysis) line endings on Windows. It suggests two fixes:

Who is C? Or is it Claude?

This branch is on a reduced github workflow. It seems that the parameters are parsed identically. (see details below). In the current version, all `\r` are filtered out anyways.

What definitely **is** different is the order in which the files are investigated. I added a sort routine to make this order deterministic.

However, for some reason there are now again 8 merge conflicts even though I merged only last week; that is why the tests do not run. I do not know if I will find the time to do this merge in the next days. :(

<details>
<summary>Grep results showing that parameters are identically parsed.</summary>
AutoOnly2 fails on Windows, not on ubuntu.

```
linux:2026-06-27T06:25:25.3047846Z     !!! Parameters for /home/runner/work/key/key/key.core/build/resources/test/de/uka/ilkd/key/scripts/jml/AutoOnly2.java: Parameters{shouldClose=false, method='null', exception='null', deleteTmpDir=true, settings=null}
windows:2026-06-27T06:26:33.9999167Z     !!! Parameters for D:\a\key\key\key.core\build\resources\test\de\uka\ilkd\key\scripts\jml\AutoOnly2.java: Parameters{shouldClose=false, method='null', exception='null', deleteTmpDir=true, settings=null}
```

</details>

mattulbrich committed:
Merge commit 'db471336507c9337d04879a4c13830afbdeac0fe' into jmlScripts

# Conflicts:
#	key.core/src/main/java/de/uka/ilkd/key/java/loader/JP2KeYConverter.java
#	key.core/src/main/java/de/uka/ilkd/key/scripts/ProofScriptEngine.java
#	key.ncore/src/main/antlr/KeYCommonParser.g4

mattulbrich committed:
Merge commit 'b83c89af98b5df800369ec523e4159d7f2172a97' into jmlScripts

# Conflicts:
#	key.core/src/main/java/de/uka/ilkd/key/scripts/ProofScriptEngine.java

mattulbrich committed:
Deprecating unused legacy script parser classes

unp1 commented:
> > > I have no idea why the Windows test cases fail. I even used my Windows PC at home to run the tests: And they succeed. ...
> 
> Thanks for investigating, Richard @unp1
> 
> > Apparently (at least C's analysis) line endings on Windows. It suggests two fixes:
> 
> Who is C? Or is it Claude?
yupp, at the moment I have still the license (not for long anzmore :-)
> 
> This branch is on a reduced github workflow. It seems that the parameters are parsed identically. (see details below). In the current version, all `\r` are filtered out anyways.
> 
> What definitely **is** different is the order in which the files are investigated. I added a sort routine to make this order deterministic.
> 
> However, for some reason there are now again 8 merge conflicts even though I merged only last week; that is why the tests do not run. I do not know if I will find the time to do this merge in the next days. :(
> 
> Grep results showing that parameters are identically parsed.

If you want I can try to get Claude doing the resolution and we both check afterwards that everything is correct?
Only problem is I don't have access to a Windows machine, so tests will run on CI.

mattulbrich commented:
If C could do the merge and you review it, that would be splendid.
The history on this branch is a bit broken since I modified the github workflows for quicker testing. So history rewrite would be required here. I was able to partially merge intermediate masters into the branch, but it is rather painful.

My bet and hope is that c0a7ae8e5feb025451008c6cbe021f0377eb2bcb will solve the problem. ..

mattulbrich commented:
> Only problem is I don't have access to a Windows machine, so tests will run on CI.

I do have a Windows machine: And it goes through there :face_with_open_eyes_and_hand_over_mouth: . But not on the CI machine. Therefore, I assume the indeterministic file order may be the problem.

This of course raises another issue: Why does this order matter? At least on could investigate then. ... if that is the reason.

unp1 committed:
Merge branch 'main' into jmlScripts

Resolve conflicts from the source-location move to key.ncore (#3870):
Position/Location now live in org.key_project.util.parsing. Updated the
stale imports (JP2KeYConverter, KeyAst, ScriptCommandAst, JmlFacade,
JMLTransformer, ApplyScriptsMacro) and moved the branch's
Location.fromPositionInfo helper into key.core's JavaSourceLocations (it
uses PositionInfo, so it cannot live in the now JavaParser-free ncore
Location).

unp1 committed:
Ignore '\r' in ScriptLineParser so CRLF line endings don't corrupt scripts

The character loop's `case '\r' -> { }` did nothing, so a carriage return fell
through to the `cmdBuilder.append(...)` below and ended up inside the command --
e.g. in a quoted value that spans a CRLF line break. Script files are marked
`text` in .gitattributes and are therefore checked out with CRLF on Windows (and
depending on the CI runner's eol config), so scripts parsed there carried stray
'\r's and could fail to match. Skip the carriage return with `continue`.

Regression test (ScriptLineParser-level) added in ProofScriptParserTest; it fails
without the fix and passes with it.

Created with AI tooling

wadoon commented:
> I do have a Windows machine: And it goes through there 🫢 . 

Windows != Windows. On Github, it is a Windows Server (2023 or 2025, I think.) 
Behavior on these machines can slightly differ.

In the old days, e.g., JVM decided for `-server` mode on Windows Server resulting into different behavior on volatile fields.

unp1 committed:
Merge branch 'main' into jmlScripts

unp1 committed:
Load admitted symbex rules independent of line endings

SymbolicExecutionOnlyMacro read its admitted rule / rule-set lists from the
bundled .txt resources and split them on "\n". On a Windows checkout these
resources have CRLF line endings (.gitattributes marks them as text), so every
entry kept a trailing '\r' ("assignment\r", "simplify_prog\r", ...).

isAdmittedRule compares the actual rule name ("assignment") against those
entries via List.contains, which then never matched. As a result no symbolic-
execution rule was admitted, the macro applied nothing, ApplyScriptsMacro never
reached a goal carrying a JmlAssert, and all goals fell through to the
TryCloseMacro fallback -- making JmlScriptTest cases close (AutoOnly2) or fail
to close (NestedAssert, ObtainFromGoal) only on Windows.

Read the lists via String.lines() and strip each entry so the names match
regardless of the checkout's line endings. Reproduced locally by converting the
two resources to CRLF (3 cases fail), fixed by this change (all 7 pass).

with AI tooling support

unp1 commented:
> I do have a Windows machine: And it goes through there 🫢 . But not on the CI machine. Therefore, I assume the indeterministic file order may be the problem.
> 
> This of course raises another issue: Why does this order matter? At least on could investigate then. ... if that is the reason.

The reason is actually quite interesting as a reminder of the subtleties between the platforms:

The problem was in SymbolicExecutionMacro: 

 ```
ADMITTED_RULES = Arrays.asList(
                Streams.toString(SymbolicExecutionOnlyMacro.class
                        .getResourceAsStream("SymbolicExecutionOnlyMacro.admittedRules.txt"))
                        .split("\n"));
ADMITTED_RULE_SETS = Arrays.asList(
                Streams.toString(SymbolicExecutionOnlyMacro.class
                        .getResourceAsStream("SymbolicExecutionOnlyMacro.admittedRuleSets.txt"))
                        .split("\n"));
```

depending on the line endings "simplify\n" --> "simplify" and a later equals check to "simplify" when deciding whether to apply the taclet or not succeeded. While when the line endings in the txt file were 
"simplify\r\n" --> "simplify\r" and the later equals check with "simplify" became false and no rules were applied.

The difference between the CI and your machine is most likely the git configuration that converts / does not convert the line endings in the files. Even so, not necessary, the fix for the deprecated ScriptLineParser is also in as additional hardening.

The merge and the commit that fixes the test are now on the branch.

wadoon committed:
restore Github workflows
